spark/src/main/scala/com/gu/tableversions/spark/SparkHiveMetastore.scala

```diff
@@ -0,0 +1,183 @@
+package com.gu.tableversions.spark
+
+import java.net.URI
+
+import cats.effect.Sync
+import cats.implicits._
+import com.gu.tableversions.core.Partition.{ColumnValue, PartitionColumn}
+import com.gu.tableversions.core._
+import com.gu.tableversions.metastore.Metastore.TableOperation
+import com.gu.tableversions.metastore.Metastore.TableOperation._
+import com.gu.tableversions.metastore.{Metastore, VersionPaths}
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.sql.SparkSession
+
+/**
+  * Concrete implementation of the Metastore API, using Spark and Hive APIs.
+  */
+class SparkHiveMetastore[F[_]](implicit spark: SparkSession, F: Sync[F]) extends Metastore[F] with LazyLogging {
+
+  import SparkHiveMetastore._
+  import spark.implicits._
+
+  override def currentVersion(table: TableName): F[TableVersion] = {
+
+    val partitionedTableVersion: F[TableVersion] = for {
+      partitions <- listPartitions(table)
+      partitionLocations <- partitions.map { partition =>
+        partitionLocation(table, toPartitionExpr(partition)).map(location => partition -> location)
+      }.sequence
+
+      partitionVersions: List[PartitionVersion] = partitionLocations.map {
+        case (partition, location) => PartitionVersion(parsePartition(partition), parseVersion(location))
+      }
+    } yield TableVersion(partitionVersions)
+
+    val snapshotTableVersion: F[TableVersion] = for {
+      tableLocation <- findTableLocation(table)
+      versionNumber = parseVersion(tableLocation)
+    } yield TableVersion(List(PartitionVersion(Partition.snapshotPartition, versionNumber)))
+
+    // Choose the calculation to perform based on whether we have a partitioned table or not
+    isPartitioned(table).flatMap(partitioned => if (partitioned) partitionedTableVersion else snapshotTableVersion)
+  }
+
+  override def update(table: TableName, changes: Metastore.TableChanges): F[Unit] =
+    changes.operations.map(appliedOp(table)).sequence.void
+
+  private def appliedOp(table: TableName)(operation: TableOperation): F[Unit] =
+    operation match {
+      case AddPartition(partitionVersion)                    => addPartition(table, partitionVersion)
+      case UpdatePartitionVersion(partitionVersion)          => updatePartitionVersion(table, partitionVersion)
+      case RemovePartition(partition)                        => removePartition(table, partition)
+      case UpdateTableLocation(tableLocation, versionNumber) => updateTableLocation(table, tableLocation, versionNumber)
```

Comment:
I find it confusing that the `UpdateTableLocation` operation (which is the only operation acting on the table level) is taking a version number. Isn't the concept of version only applying to a partition, not a table?

Reply:
In the case of snapshot tables, it is the main table location we have to change and not partition versions. So the question is how we model that through this API. I guess the other option is to use `UpdatePartitionVersion` for this as well, using the 'special' partition for snapshots. This may well be better, I'll try to do this and see how it looks.

Actually, that reminds me: I'm considering updating the model to not use the "special" / empty partition for snapshot tables but have ADTs with explicit types for the partitioned and versioned cases instead. That might play into how this looks too.

Reply:
As mentioned in a reply below, I think it's worth keeping the explicit operation to update a base table location, distinct from updating partitions. I did make a different change in this area though: I realised it didn't make sense to pass in the table location when updating the table version, so instead of:

```scala
case UpdateTableLocation(tableLocation, versionNumber) => updateTableLocation(table, tableLocation, versionNumber)
```

It's now:

```scala
    final case class UpdateTableVersion(versionNumber: VersionNumber) extends TableOperation
```

Let me know what you think.